src/doc/unstable-book/src/language-features/c-ffi-pure.md

```diff
@@ -0,0 +1,55 @@
+# `c_ffi_pure`
+
+The `#[c_ffi_pure]` attribute applies clang's `pure` attribute to foreign
+functions declarations. 
+
+That is, `#[c_ffi_pure]` functions shall have no effects except for its return
+value, which shall not change across two consecutive function calls and can only
+depend on the values of the function parameters and/or global memory.
```

Comment:
A `pure` function can read `static`s, including `static`s that are pointers, and read memory through these pointers.

Reply:
Yes, it can do all that and even more. And I don't think this paragraph properly conveys that.

Reply:
Do you have any suggestion? 

Being able to read caller local variables via pointers passed to them is AFAICT already covered by "depend on the values of the function parameters", and by global memory I mean any memory that is globally accessible, e.g., via an static, a magic address, etc.

Reply:
(Sorry for the delay.)

I think language like in the GCC docs (arbitrary memory) is better. The phrase "values of the function parameters" is not clear for pointer arguments because it can be read as "only depend on the pointer itself, not the contents of the memory it points to" and indeed that's sort of the case for `c_ffi_const` functions. For example, IIUC `uintptr_t tag_pointer(void *, int tag)` that does arithmetic on the address can be `c_ffi_const` and work with pointers to non-constant memory, but of course reading from the pointed-to memory would generally make it non-`c_ffi_const`.